File: CoreML-YOLOv8 Pork/CoreML-YOLOv8 Pork-OC/CoreML- YOLOv8 Pork-OC/ViewController.m
//
//  ViewController.m
//  CoreML- YOLOv8 Pork-OC
//
//  Created by 杨帆 on 2024/5/12.
//

#import "ViewController.h"

@interface ViewController ()

@property (nonatomic) CGSize bufferSize;
@property (nonatomic) AVCaptureSession *session;
@property (nonatomic) AVCaptureVideoPreviewLayer *previewLayer;
@property (nonatomic) AVCaptureVideoDataOutput *videoDataOutput;
@property (nonatomic) NSMutableArray<VNRequest *> *requests;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)setup {
    [self setupAVCapture];
    [self setupLayers];
    [self setupModels];
    [self startCaptureSession];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self setup];
}

- (void)startCaptureSession {
    [self.session startRunning];
}

- (void)teardownAVCapture {
    [self.previewLayer removeFromSuperlayer];
    self.previewLayer = nil;
}

- (void)setupAVCapture {
    AVCaptureDeviceInput *deviceInput = nil;
    AVCaptureDevice *videoDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    NSError *error = nil;
    deviceInput = [AVCaptureDeviceInput deviceInputWithDevice:videoDevice error:&error];
    if (error) {
        NSLog(@"Could not create video device input: %@", error);
        return;
    }
    
    self.session = [[AVCaptureSession alloc] init];
    self.session.sessionPreset = AVCaptureSessionPreset640x480;
    
    if ([self.session canAddInput:deviceInput]) {
        [self.session addInput:deviceInput];
    } else {
        NSLog(@"Could not add video device input to the session");
        return;
    }
    
    self.videoDataOutput = [[AVCaptureVideoDataOutput alloc] init];
    if ([self.session canAddOutput:self.videoDataOutput]) {
        [self.session addOutput:self.videoDataOutput];
        self.videoDataOutput.alwaysDiscardsLateVideoFrames = YES;
        NSDictionary *videoSettings = @{(NSString *)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_420YpCbCr8BiPlanarFullRange)};
        [self.videoDataOutput setVideoSettings:videoSettings];
        dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
        [self.videoDataOutput setSampleBufferDelegate:self queue:queue];
    } else {
        NSLog(@"Could not add video data output to the session");
        return;
    }
    
    AVCaptureConnection *captureConnection = [self.videoDataOutput connectionWithMediaType:AVMediaTypeVideo];
    captureConnection.enabled = YES;
    
    [videoDevice lockForConfiguration:nil];
    CMVideoDimensions dimensions = CMVideoFormatDescriptionGetDimensions(videoDevice.activeFormat.formatDescription);
    self.bufferSize = CGSizeMake(dimensions.width, dimensions.height);
    [videoDevice unlockForConfiguration];
    
    [self.session commitConfiguration];
}

- (void)setupLayers {
    self.previewLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.session];
    self.previewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    self.previewLayer.frame = self.view.layer.bounds;
    [self.view.layer addSublayer:self.previewLayer];
}

- (void)setupModels {
    NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"best100" withExtension:@"mlmodelc"];
    if (!modelURL) {
        return;
    }
    
    VNCoreMLModel *visionModel = nil;
    NSError *error = nil;
    MLModel *mlModel = [MLModel modelWithContentsOfURL:modelURL error:&error];
    if (!error) {
        visionModel = [VNCoreMLModel modelForMLModel:mlModel error:&error];
    }
    
    if (error) {
        NSLog(@"Model loading went wrong: %@", error);
        return;
    }
    
    VNCoreMLRequest *objectRecognition = [[VNCoreMLRequest alloc] initWithModel:visionModel completionHandler:^(VNRequest * _Nonnull request, NSError * _Nullable error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            NSArray *results = request.results;
            if (results) {
                [self drawVisionRequestResults:results];
            }
        });
    }];
    
    self.requests = [NSMutableArray arrayWithObject:objectRecognition];
}

- (void)drawVisionRequestResults:(NSArray *)results {
    for (id observation in results) {
        if ([observation isKindOfClass:[VNRecognizedObjectObservation class]]) {
            VNRecognizedObjectObservation *objectObservation = (VNRecognizedObjectObservation *)observation;
            VNClassificationObservation *topLabelObservation = objectObservation.labels.firstObject;
            CGRect objectBounds = VNImageRectForNormalizedRect(objectObservation.boundingBox, (int)self.bufferSize.width, (int)self.bufferSize.height);
            NSLog(@"位置信息： %@", NSStringFromCGRect(objectBounds));
            NSLog(@"标签： %@", topLabelObservation.identifier);
            NSLog(@"置信度： %f", topLabelObservation.confidence);
        }
    }
}

@end

@implementation ViewController (Utilities)

- (CGImagePropertyOrientation)exifOrientationFromDeviceOrientation {
    UIDeviceOrientation curDeviceOrientation = UIDevice.currentDevice.orientation;
    CGImagePropertyOrientation exifOrientation;
    switch (curDeviceOrientation) {
        case UIDeviceOrientationPortraitUpsideDown:
            exifOrientation = kCGImagePropertyOrientationLeft;
            break;
        case UIDeviceOrientationLandscapeLeft:
            exifOrientation = kCGImagePropertyOrientationUpMirrored;
            break;
        case UIDeviceOrientationLandscapeRight:
            exifOrientation = kCGImagePropertyOrientationDown;
            break;
        case UIDeviceOrientationPortrait:
            exifOrientation = kCGImagePropertyOrientationUp;
            break;
        default:
            exifOrientation = kCGImagePropertyOrientationUp;
            break;
    }
    return exifOrientation;
}


- (void)captureOutput:(AVCaptureOutput *)output didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
    CVPixelBufferRef pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
    if (!pixelBuffer) {
        return;
    }
    
    CGImagePropertyOrientation exifOrientation = [self exifOrientationFromDeviceOrientation];
    VNImageRequestHandler *imageRequestHandler = [[VNImageRequestHandler alloc] initWithCVPixelBuffer:pixelBuffer orientation:exifOrientation options:@{}];
    NSError *error = nil;
    [imageRequestHandler performRequests:self.requests error:&error];
    if (error) {
        NSLog(@"%@", error);
    }
}


@end

File: CoreML-YOLOv8 Pork/CoreML-YOLOv8 Pork-Swift/CoreML- YOLOv8 Pork/ViewController.swift
//
//  ViewController.swift
//  CoreML- YOLOv8 Pork
//
//  Created by 杨帆 on 2024/5/11.
//

import AVFoundation
import UIKit
import Vision

class ViewController: UIViewController {
    private var bufferSize: CGSize = .zero
    private let session = AVCaptureSession()
    private var previewLayer: AVCaptureVideoPreviewLayer!
    private let videoDataOutput = AVCaptureVideoDataOutput()
    private var requests = [VNRequest]()

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    func setup() {
        setupAVCapture()
        setupLayers()
        setupModels()
        startCaptureSession()
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        setup()
    }

    func startCaptureSession() {
        session.startRunning()
    }

    func teardownAVCapture() {
        previewLayer.removeFromSuperlayer()
        previewLayer = nil
    }
}

extension ViewController {
    func setupAVCapture() {
        var deviceInput: AVCaptureDeviceInput!
        let videoDevice = AVCaptureDevice.DiscoverySession(deviceTypes: [.builtInWideAngleCamera], mediaType: .video, position: .back).devices.first
        do {
            deviceInput = try AVCaptureDeviceInput(device: videoDevice!)
        } catch {
            print("Could not create video device input: \(error)")
            return
        }
        session.beginConfiguration()
        session.sessionPreset = .vga640x480
        guard session.canAddInput(deviceInput) else {
            print("Could not add video device input to the session")
            session.commitConfiguration()
            return
        }
        session.addInput(deviceInput)
        if session.canAddOutput(videoDataOutput) {
            session.addOutput(videoDataOutput)
            videoDataOutput.alwaysDiscardsLateVideoFrames = true
            videoDataOutput.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: Int(kCVPixelFormatType_420YpCbCr8BiPlanarFullRange)]
            videoDataOutput.setSampleBufferDelegate(self, queue: DispatchQueue.global())
        } else {
            print("Could not add video data output to the session")
            session.commitConfiguration()
            return
        }
        let captureConnection = videoDataOutput.connection(with: .video)
        captureConnection?.isEnabled = true
        do {
            try videoDevice!.lockForConfiguration()
            let dimensions = CMVideoFormatDescriptionGetDimensions((videoDevice?.activeFormat.formatDescription)!)
            bufferSize.width = CGFloat(dimensions.width)
            bufferSize.height = CGFloat(dimensions.height)
            videoDevice!.unlockForConfiguration()
        } catch {
            print(error)
        }
        session.commitConfiguration()
    }

    func setupLayers() {
        previewLayer = AVCaptureVideoPreviewLayer(session: session)
        previewLayer.videoGravity = AVLayerVideoGravity.resizeAspectFill
        previewLayer.frame = view.layer.bounds
        view.layer.addSublayer(previewLayer)
    }
}

extension ViewController {
    func setupModels() {
        guard let modelURL = Bundle.main.url(forResource: "best100", withExtension: "mlmodelc") else {
            return
        }
        do {
            let visionModel = try VNCoreMLModel(for: MLModel(contentsOf: modelURL))
            let objectRecognition = VNCoreMLRequest(model: visionModel) { request, _ in
                DispatchQueue.main.async {
                    if let results = request.results {
                        self.drawVisionRequestResults(results)
                    }
                }
            }
            requests = [objectRecognition]
        } catch {
            print("Model loading went wrong: \(error)")
        }
    }

    func drawVisionRequestResults(_ results: [Any]) {
        for observation in results where observation is VNRecognizedObjectObservation {
            guard let objectObservation = observation as? VNRecognizedObjectObservation else {
                continue
            }

            let topLabelObservation = objectObservation.labels[0]
            let objectBounds = VNImageRectForNormalizedRect(objectObservation.boundingBox, Int(bufferSize.width), Int(bufferSize.height))
            print("置信度：", topLabelObservation.confidence)
            print("内容：", topLabelObservation.identifier)
            print("边框", objectBounds)
        }
    }

    func exifOrientationFromDeviceOrientation() -> CGImagePropertyOrientation {
        let curDeviceOrientation = UIDevice.current.orientation
        let exifOrientation: CGImagePropertyOrientation
        switch curDeviceOrientation {
        case UIDeviceOrientation.portraitUpsideDown:
            exifOrientation = .left
        case UIDeviceOrientation.landscapeLeft:
            exifOrientation = .upMirrored
        case UIDeviceOrientation.landscapeRight:
            exifOrientation = .down
        case UIDeviceOrientation.portrait:
            exifOrientation = .up
        default:
            exifOrientation = .up
        }
        return exifOrientation
    }
}

extension ViewController: AVCaptureVideoDataOutputSampleBufferDelegate {
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            return
        }
        let exifOrientation = exifOrientationFromDeviceOrientation()
        let imageRequestHandler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, orientation: exifOrientation, options: [:])
        do {
            try imageRequestHandler.perform(requests)
        } catch {
            print(error)
        }
    }
}


